### DISCUSSION TITLE
DB관련 yml 설정파일 논의

### DISCUSSION THREAD

**Discussion by dongho108** [woowacourse-teams/2022-ternoko#63] [Show and tell]
[https://github.com/woowacourse-teams/2022-ternoko/pull/62](https://github.com/woowacourse-teams/2022-ternoko/pull/62)

위의 PR 관련해서 논의드리려고 합니다!

현재 default를 dev로 설정해놓고 있는데, local에서 실행할때는 `README.md` 에 적힌 설정을 해주어야 실행이 됩니다.

ddl:create가 default인 것이 안전하지 않다고 생각해서 dev를 default로 설정을 했는데요.
그러다보니 로컬에선 개발자가 직접 설정을 해주어야만 돌아가는 상황이 생깁니다.

편의성과 안정성에대해 trade-off를 생각해봤을 때,
production 배포 서버에서 `ddl:create`가 되는 상황을 `최대한 막고` 배포서버에서 실행시킬때 옵션을 넣어 실행시키는 것이 가장 최선의 방법이라고 생각했습니다.

아래는 yml 파일 설정 설계입니다.

| yml | DB | DB 위치 | ddl | 유저권한 | id/password | 포트번호 |
| --- | --- | --- | --- | --- | --- | --- |
| test | h2 (메모리) | localhost | create-drop | * | sa /  | . |
| local | mysql | localhost | create | * | root / root | 3306 |
| dev | mysql | test-DB서버 | validate | select,update,delete | ternoko / ternoko1 | 53306 |
| prod | mysql | prod-DB서버 | validate | select,update,delete | ternoko / {복잡한password} | 53306 |

### default를 dev로 설정시

- 장점 : (안정성) prod나 dev의 데이터가 삭제될(`ddl:create`) 일은 거의없다.
- 단점 : local에서 clone을 받을때마다 active profiles 를 설정해주어야 한다.

### default를 local로 설정시

- 장점 : local에서는 개발자가 따로 설정할 필요가 없다.
- 단점 :
    1. `ddl:create`가 default여서 dev나 prod에서 데이터가 삭제될 가능성이 생긴다.
        - 해결방법 : 여러개의 방어막을 둔다.
            1. dev와 prod에서는 DB 서버 분리
            2. 유저 정보 다르게 설정
            3. DB 유저 권한 create, drop 제거
            4. 포트번호 다르게 설정
    2. dev서버나 prod서버에서 실행시 active profiles를 지정해서 실행시켜야 한다.
        - 해결방법 : 스크립트로 자동화한다.
    

어떤 방식이 좋을지 논의하고싶어요!

**Comment by HyeonbinSa** [comment]
우선 배포 관련해서 계속해서 신경써주셔서 감사드립니다.
저는 dev로 하는 방향으로 갔으면 좋겠습니다.
아직은 데모 기간으로 데이터의 유실이 생기더라도 문제가 되지 않지만 추후 운영 서버로 사용한다고 생각했을 때 데이터의 삭제가 일어나면 안될 것 같습니다.
개발할 때 불편함이 있더라도 안정적인 서비스 운영이 우선시되어야할 것 같습니다.
관련해서 블로그 [포스팅](https://smpark1020.tistory.com/m/140)이 있어 공유드립니다.

추후에 더 나은 방안이 있는지 찾아보도록 하겠습니다.

**Reply by dongho108** [reply]
좋은 의견 감사합니다. 개발이나 운영서버에서 create를 실행시키자는 말은 아니었는데 ddl 옵션만 보면 희박한 확률로 실행이 될 수도 있을 것 같아요.

**Reply by HyeonbinSa** [reply]
희박한 확률에서 발생하는 상황이 어떤게 있을지 예로 설명해주실 수 있을까요??

**Reply by dongho108** [reply]
아래 열음 댓글에 상황을 한번 달아봤습니다!

**Comment by Juhyung990122** [comment]
dev와 prod에서는 Mysql을, local 에서는 H2을 사용하고 있으므로 DB가 이미 분리되어있고 분리체제 그대로 가는게 맞는 것 같습니다.

이에 현재 dev의 ddl설정은 MYSql의 update로 설정되어있고 local의 ddl설정은 H2의 create로 설정되어있으므로 우려하시는 상황은 저희중 누군가 dev.yml의 설정값을 create로 바꾸지 않는 한 일어나지 않습니다. 그럼에도 불구하고 우려되신다면 dev 프로필은 소스코드가 아닌 서비스 서버내에 두어 개발자가 쉽게 수정하지 못하게 하는것도 방법일 것 같네요!

또한 혹시나 실수로 서버 내의 deploy.sh이나 java -jar 에 dev 옵션을 주지 않더라도 이미 local과 dev가 서로 가리키는 DB가 다르므로 실제 사용되는 데이터가 삭제되지 않을 것 이라 생각됩니다.

이에더해 보편적으로 local이 디폴트값인지라 Git에서 우리의 체제(dev가 디폴트)를 모르는 사람이 클론을 받아 프로젝트를 실행시키기 위해서는 Readme.me를 필수적으로 읽고 ide 실행전에 적용시켜야합니다. 해당 설정이 누락되면 프로젝트가 정상적으로 실행되지 않을테니까요🤔 다른 개발자가 코드 실행을 위해 추가적으로 파악해야하는 점이 있다는점이 마음에 걸리네요.

따라서 저는 local을 디폴트로 가는게 좋은 것 같아요. 또한 local을 디폴트로 가게된다면 제안주신 방어체제 적용하면 안정성또한 확보할 수 있을 것 같아 좋습니다☺️ 정리 감사하고 고민하시느라 고생많으셨어요!

**Reply by dongho108** [reply]
`local이 디폴트값인지라 Git에서 우리의 체제(dev가 디폴트)를 모르는 사람이 클론을 받아 프로젝트를 실행시키기 위해서는 Readme.me를 필수적으로 읽고 ide 실행전에 적용시켜야합니다. ` 는 사실 당연한거라서 걱정은 되지않아요!
사실 개발자 클릭세번이면 안정성을 확보할 수 있는 방법이있는데 굳이 여러단계 방어막을 거치면서 안정성이 조금이라도 떨어지는 방법을 사용하는 방법이 마음에 걸리긴 하네요😂

**Reply by Juhyung990122** [reply]
좋습니다. 다만 안정성 측면에서도 저는 좀 생각이 다른게, 앞서 말씀드린 것 처럼 두 설정파일은 이미 서로 다른 데이터베이스를 보고있는데 실수로라도 local이 적용되었을때 dev와 prod가 보고있는 데이터베이스 내의 데이터들이 삭제될 일이 있을까요? 구체적으로 상황을 들어주시면 감사하겠습니다🙂

**Reply by dongho108** [reply]
물론 지금 파일 그대로 (4단계의 방어단계를 넣은상태로) 실행시킨다면 dev, prod DB 의 데이터가 drop-create 될 확률은 없다고 볼 수 있습니다.

데이터 관련해서는 항상 만약을 가정하기 때문에 상황을 가정해본다면,
어느 개발자는 local.yml에서 datasource 관련 설정을 dev나 prod의 mysql설정으로 바꿀 수도 있습니다.
본인은 "dev나 prod의 ddl을 바꾸지 않았기 때문에 dev, prod DB에 절대로 영향을 미치지 않겠다"고 생각하겠죠

"만약 default가 dev라면 local을 아무리 수정해도 dev, prod에 아무영향을 미칠 수 없겠지만, default가 local이라면 local을 수정했을때 실제 DB를 건드릴 수 있겠다"고 생각했습니다.

**Reply by dongho108** [reply]
그런데 이 상황도 사실 미리생각했어서, DB 유저권한에 create, drop을 막아놨습니다..ㅋㅋ! 😁

**Reply by Juhyung990122** [reply]
음 dev와 prod 내의 데이터소스 접근을 위한 값(username과 password)이 단순 스트링으로 관리된다면 말씀주신 상황이 발생할 여지가 있으나 현재 실제 사용되는 DB 정보는 ec2서버 내의 환경변수로 관리되고 있습니다. 이 환경 변수를 알아내어 local.yml 에 넣기 위해선 아시다싶이 ssh 로 어플리케이션 서버에 접근해야하며, 이를 위해선 pem키가 노출되어야하는데 그쯤 됐으면 더이상 DB 관리를 넘어서 보안측면의 치명적인 문제가 생긴걸로 봐야할 것 같네요. 

각자의 로컬에서 사용하는 pem키가 노출될 확률과 앤지께서 말씀해주신 수정비용을 따져보았을때 저는 전자가 더 발생확률이 낮다고 생각하기 때문에 local 을 디폴트로 가는것에 한표 넣고 의견 마무리하겠습니다! 감사합니다.

**Reply by dongho108** [reply]
적극적인 의견 표출 감사합니다!!!!!! 👍👍 👍

**Comment by soominsohn** [comment]
저는 `default를 local로 설정`에 한표 하겠습니다!
데이터가 삭제될 단점이 있지만, 이를 막을 수 있는 해결책이 충분하다고 생각합니다.
적어주신 대로 데이터 유실을 막을 수 있는 방어막도 있고, 스크립트로 active profiles를 실행시켜 자동화 또한 가능하므로 local설정으로 가도 좋을 것 같습니다. 
배포할 때 한번 더 체크하는 시간적 비용보다, 로컬에서 default가 local(create)로 되어있지 않아 생기는 버그를 잡는 시간적 비용(테이블에 데이터 중복으로 들어가서 생기는 문제 등)이 더 클 것이라고 생각이 드네요 😄

**Reply by dongho108** [reply]
1. 로컬에서 스크립트를 만드는 비용보다 edit configurations 설정이 더 적다고 생각이 듭니다!
2. 로컬에서 default (dev) 를 실행시킨다면 mysql은 ddl:create되지 않아 데이터가 중복으로 들어갈 일은 없고 실행이 안되거나, 잘되거나 둘중 하나일거라고 생각이 드네요
